File: 28-03-2024/City With the Smallest Number of Neighbors at a Threshold Distance.js
//{ Driver Code Starts
//Initial Template for javascript

'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => { inputString += inputStdin; });

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(
        string => { return string.trim(); });

    main();
});

function readLine() {
    return inputString[currentLine++];
}


function main() {
    let t = parseInt(readLine());
    let i = 0;
    for (; i < t; i++) {
        let [n, m] = readLine().trim().split(" ").map((x) => parseInt(x));
        let edges = [];
        for (let i = 0; i < m; i++) {
            let inputLine = readLine().trim().split(" ").map((x) => parseInt(x));
            edges.push(inputLine);
        }
        let d = parseInt(readLine());
        let obj = new Solution();
        let res = obj.findCity(n, m, edges, d);
        console.log(res);
    }
}
// } Driver Code Ends


//User function Template for JavaScript
/**
 * @param {number} n
 * @param {number} m
 * @param {number[][]} edges
 * @param {number} distanceThreshold
 * @returns {number}
*/

class Solution {
    //Function to find the city with the smallest number of neighbors
    findCity(n, m, edges, distanceThreshold) {
        //Your Code Start
        const g = new Map();
        
        for (const [u, v, w] of edges) {
            if (!g.has(u)) g.set(u, []);
            if (!g.has(v)) g.set(v, []);
            g.get(u).push([v, w]);
            g.get(v).push([u, w]);
        }
        
        function dijkstra(n) {
            const costs = new Map([[n, 0]]);
            const q = [[0, n]];
            while (q.length > 0) {
                const [cost0, n0] = q.shift();
                for (const [nbr, c] of g.get(n0) || []) {
                    const cost = cost0 + c;
                    if (cost > distanceThreshold) continue;
                    if (!costs.has(nbr) || costs.get(nbr) > cost) {
                        costs.set(nbr, cost);
                        q.push([cost, nbr]);
                    }
                }
            }
            return costs.size;
        }
        
        let ans = 0;
        let cnt = n;
        for (let k = 0; k < n; k++) {
            const c = dijkstra(k);
            if (c <= cnt) {
                cnt = c;
                ans = k;
            }
        }
        return ans;
 //Your Code End
    }
}

File: 28-03-2024/City With the Smallest Number of Neighbors at a Threshold Distance.py
#User function Template for python3

from typing import List
#Your Code Start
class Solution:
    def findCity(self, n : int, m : int, edges : List[List[int]], d : int) -> int:
        # code here
        from collections import defaultdict
        from heapq import heappush, heappop
        g = defaultdict(list)
        
        for u, v, w in edges:
            g[u].append((v, w))
            g[v].append((u, w))
        
        def dijkstr(n):
            costs = {n: 0}
            q = [(0, n)]
            while q:
                cost0, n0 = heappop(q)
                for nbr, c in g[n0]:
                    cost = cost0+c
                    if cost > d:
                        continue
                    if nbr not in costs or costs[nbr] > cost:
                        costs[nbr] = cost
                        heappush(q, (cost, nbr))
            return len(costs)
            
        ans = 0
        cnt = n
        for k in range(n):
            c = dijkstr(k)
            #print(f"k = {k}, c = {c}")
            if c <= cnt:
                cnt = c
                ans = k
        return ans
#Your Code End


#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    T = int(input())
    for t in range(T):
        
        n, m = map(int, input().strip().split())
        edges = []
        for i in range(m):
            u, v, wt = map(int, input().strip().split())
            edges.append([u, v, wt])
        distanceThreshold = int(input())
        obj = Solution()
        ans = obj.findCity(n, m, edges, distanceThreshold)
        print(ans)
            

# } Driver Code Ends
